libs/ojsonnet/Core_jsonnet.ml

```diff
@@ -161,11 +161,65 @@ and for_comp = tok (* 'for' *) * ident * tok (* 'in' *) * expr
 [@@deriving show { with_path = false }]
 
 (*****************************************************************************)
-(* Program *)
+(* Env *)
 (*****************************************************************************)
+type env = {
+  (* The spec uses a lambda-calculus inspired substitution model, but
+   * it is probably simpler and more efficient to use a classic
+   * environment where the locals are defined. Jsonnet uses lazy
+   * evaluation so we model this by allowing unevaluated expressions in environment below.
+   *)
+  locals : (local_id, local) Map_.t;
+  (* for call tracing *)
+  depth : int;
+}
 
-type program = expr [@@deriving show]
+and local_id = LSelf | LSuper | LId of string
+and local = Val of (value_ * env) | Expr of (expr * env)
```

Comment:
you could rename this type to lazy_expr or even better lazy_value and add a comment about the environment, that it's used to be simply value_ Lazy.t but that we switch to explicit env because of + and the need to access the environment etc. etc. (and that with lazy/closure there was no way we could do that).

Reply:
Maybe it could be  Val of ... | Unevaluated of expr * env
and also because env needs to be present in both case, we could use a record instead and
document why we also need env even for the Val case (which is subtle, but useful because of nested
array elts or fields, and again this is the kind of things that is nice to document in comments).